File: Algorithms/Binary Search.cpp
#include<bits/stdc++.h>
using namespace std;

/*Binary Search algorithm searches for an element in a sorted array
by repeatedly dividing the array into two halves.*/


int binarySearch(int arr[], int left, int right, int element)
{
	if(right>=left)
	{
		int mid = left + (right-left)/2; 

		//If element is present at the middle return mid
		if(arr[mid] == element) 						
			return mid;

		//If element is smaller than middle element then search for the element in left subarray
		if(arr[mid] > element)                 				
			return binarySearch(arr,left,mid-1,element);

		//Else search in right subarray
		return binarySearch(arr,mid+1,right,element);

	}
	//If element is not found then return -1
	return -1;
}

int main()
{
	int size;
	cin>>size;
	int arr[size];
	for(int i=0;i<size;i++){
		cin>>a[i];
	}
	int element = 12;

	int result = binarySearch(arr,0,size-1,element);
	if(result==-1)
		cout<<"Element is not present in array"<<endl;
	else
		cout<<"Element is present at index "<<result<<endl;

	return 0;
}

File: Algorithms/JAVA/Binary_Search.java

import java.util.*;
import java.io.IOException;

class Binary_Search{

    static int binarySearch(int arr[], int start, int end, int key){
        
        if(end>=start)
        {
            int mid= start + (end-start)/2;

            if(arr[mid]==key)
                return mid;

            else if(key<arr[mid])
                return binarySearch(arr, start, mid-1, key);

            return binarySearch(arr, mid+1, end, key);
        }
        return -1;
    }

    public static void main(String[] args)throws IOException{
    Scanner sc=new Scanner(System.in);
    int n = sc.nextInt();
    int key=sc.nextInt();
    int[] arr=new int[n];
    for(int i=0;i<n;i++)
        arr[i]=sc.nextInt();
    int ans=binarySearch(arr, 0, n-1, key);
    if(ans==-1)
        System.out.println("Element is not present in the Array");
    else    
        System.out.println("Element is present at index "+ans);
    sc.close();
    }
}